### PR TITLE
Give dynamically generated instructions on how to replicate x.py errors

### LINKED ISSUE #84742: Suggest how to recreate a build failure with the fewest possible steps
Tidy (and all tests really) should suggest how to replicate exactly what went wrong without running a full x.py test. That helps people fix their mistakes much more quickly without waiting for a full build.

_Originally posted by @jyn514 in https://github.com/rust-lang/rust/issues/75987#issuecomment-829967515_

### PR DESCRIPTION
**Pull request by jyn514**
Example output:

```
Checking stage0 rustdoc artifacts (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu)
    Checking rustdoc v0.0.0 (/home/joshua/rustc/src/librustdoc)
error: expected item, found `/`
 --> src/librustdoc/lib.rs:1:1
  |
1 | /
  | ^ expected item

error: aborting due to previous error

error: could not compile `rustdoc`

To learn more, run the command again with --verbose.
command did not execute successfully: "/home/joshua/rustc/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "check" "--target" "x86_64-unknown-linux-gnu" "-Zbinary-dep-depinfo" "-j" "8" "--release" "--manifest-path" "/home/joshua/rustc/src/tools/rustdoc/Cargo.toml" "--message-format" "json-render-diagnostics"
expected success, got: exit status: 101
note: failed while building bootstrap::check::Rustdoc
help: to replicate this failure, run `./x.py check src/librustdoc`
failed to run: /home/joshua/rustc/build/bootstrap/debug/bootstrap check src/librustdoc
Build completed unsuccessfully in 0:00:01
```

Happy to take suggestions on how to improve the error message :) right
now it's getting lost in all the other output. Maybe it makes sense to
use bold and colors?

This works by storing the current step at all times in a global Mutex
and registering an `at_exit` handler which looks up the current step.

Fixes https://github.com/rust-lang/rust/issues/84742.

### CODE REVIEW

**Comment by jyn514** [comment]
image

**Comment by jyn514** [comment]
Hmm, this doesn't work very well for unit tests:
```
failures:
    test_decode_option_none

test result: FAILED. 69 passed; 1 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.01s

error: test failed, to rerun pass '-p rustc_serialize --test json'


command did not execute successfully: "/home/joshua/rustc/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "test" "--target" "x86_64-unknown-linux-gnu" "-Zbinary-dep-depinfo" "-j" "8" "--release" "--features" " llvm" "--manifest-path" "/home/joshua/rustc/compiler/rustc/Cargo.toml" "-p" "rustc_serialize" "--" "--quiet"
expected success, got: exit status: 101


note: failed while building bootstrap::test::Crate
help: to replicate this failure, run `./x.py test library/alloc`
```

**Comment by jyn514** [comment]
Ok, should be better now:
```
note: failed while building bootstrap::test::Crate
help: to replicate this failure, run `./x.py test rustc_serialize --stage 0`
```

**Comment by jyn514** [comment]
UI tests are not super helpful though:
```
failures:
    [ui] ui/tls.rs

test result: FAILED. 0 passed; 1 failed; 1 ignored; 0 measured; 11954 filtered out; finished in 0.03s

note: failed while building bootstrap::test::Compiletest
help: to replicate this failure, run `./x.py test /home/joshua/rustc --stage 1`
```

File: src/bootstrap/flags.rs [hunk 1]
```diff
@@ -27,6 +29,16 @@ impl Default for Color {
     }
 }
 
+impl From<Color> for ColorChoice {
```
Comment (jyn514) [inline src/bootstrap/flags.rs:32]: Since I'm adding termcolor anyway, it probably makes sense to remove Color and just use ColorChoice directly.

**Comment by Mark-Simulacrum** [comment]
FWIW I'd prefer to avoid coloring or otherwise formatting things in bootstrap. We can likely hide more of the failed command message (e.g., showing the full compile test invocation isn't necessary in non-verbose mode imo).

**Comment by jyn514** [comment]
> FWIW I'd prefer to avoid coloring or otherwise formatting things in bootstrap. We can likely hide more of the failed command message (e.g., showing the full compile test invocation isn't necessary in non-verbose mode imo).

Sure thing, done. I like it better with less output anyway :)

**Comment by Mark-Simulacrum** [comment]
I am unwilling to accept the atexit and/or similar primitives to facilitate this logic. I worry that the benefit it brings seems fairly low; in practice, the full build is likely to either take a while (if, for example, the compiler has been modified, it will need to be rebuilt) or not happen -- we already should sufficiently cache that we don't rerun tests that passed without cause.

Plus, this sets precedent for us to try and fine tune the logic to allow even more precise diagnostic reporting; I'm not currently inclined to do that.

I think an alternative to this, which may be more broadly useful, is to change the formatting of the "step start" messages we currently emit such that they contain a more close reference to the necessary command to invoke that step, should any paths be available. For example, for the errors shown in this PR, we could have a log something like this, which hopefully gives a clearer indication that to rerun you can simply copy/paste the most recent step. That info: line could even be auto generated (though perhaps not auto-emitted), similar to what you've done in this PR.

```
info: check src/librustdoc --stage 0 --host x86_64-unknown-linux-gnu --target x86_64-unknown-linux-gnu
    Checking rustdoc v0.0.0 (/home/joshua/rustc/src/librustdoc)
error: expected item, found `/`
 --> src/librustdoc/lib.rs:1:1
  |
1 | /
  | ^ expected item

error: aborting due to previous error

error: could not compile `rustdoc`

To learn more, run the command again with --verbose.
note: failed while building bootstrap::check::Rustdoc
failed to run: /home/joshua/rustc/build/bootstrap/debug/bootstrap check src/librustdoc
Build completed unsuccessfully in 0:00:01
```

**Comment by jyn514** [comment]
> I think an alternative to this, which may be more broadly useful, is to change the formatting of the "step start" messages we currently emit such that they contain a more close reference to the necessary command to invoke that step, should any paths be available.

This is a great idea, thank you!

> Plus, this sets precedent for us to try and fine tune the logic to allow even more precise diagnostic reporting; I'm not currently inclined to do that.

> That info: line could even be auto generated (though perhaps not auto-emitted), similar to what you've done in this PR.

Hmm, I'm not sure I understand how these fit together? I'm fine with not using `atexit`, but I don't see how that will mean people don't expect it to be more precise. Maybe you meant that it will be easier to fine-tune because it's at the place the step is run (that's a convincing argument to me) but I didn't get the impression you wanted to keep the reporting precise.

**Comment by jyn514** [comment]
I marked this as waiting-on-review since I'm not sure I'm on the right track - feel free to flip back to waiting-on-author if you want me to convert everything to use the new `step_info()` function.

**Comment by Mark-Simulacrum** [comment]
Just responding quickly to the last comment (haven't actually looked at code yet): changing to formatting step descriptions prior to running them in a way closely matching with their invocation is inherently high-level by virtue of steps being high level; for example, compile test invocation wouldn't/shouldn't specify a specific test in the info line, but the previous style / strategy would have been much more likely to take that tack. Doing so before the failure occurs similarly is a forcing function to print more limited information; again, this pushes the code in a less invasive direction hopefully. It also avoids dependence on primitives like atexit, which should be avoided if at all possible in my opinion.

File: src/bootstrap/builder.rs [hunk 2]
```diff
@@ -590,6 +621,29 @@ impl<'a> Builder<'a> {
...
+    /// This serves two purposes:
+    /// 1. Describe what step is currently being run.
+    /// 2. Describe how to run only this step in case it fails.
+    pub(crate) fn step_info(&self, step: &impl Step) {
+        if self.config.dry_run {
+            return;
+        }
+        print!(
+            "{} {} --stage {}",
+            // TODO: this is wrong, e.g. `check --stage 1` runs build commands first
```
Comment (Mark-Simulacrum) [inline src/bootstrap/builder.rs:635]: Not entirely sure how to fix this, but it seems critical to do so if we're to have some measure of success. I am hesitant to add additional annotations to the steps themselves, but perhaps that might be the way to go.

**Comment by jyn514** [comment]
> I think an alternative to this, which may be more broadly useful, is to change the formatting of the "step start" messages we currently emit such that they contain a more close reference to the necessary command to invoke that step, should any paths be available.

Hmm, so my concern with this is very easily gets lost in cargo's output:

<details>

```
$ x check src/librustdoc
check src/librustdoc --stage 0
   Compiling autocfg v1.0.0
    Checking lazy_static v1.4.0
   Compiling proc-macro2 v1.0.24
   Compiling libc v0.2.93
   Compiling unicode-xid v0.2.1
   Compiling syn v1.0.65
    Checking cfg-if v1.0.0
   Compiling memchr v2.4.0
    Checking cfg-if v0.1.10
    Checking regex-syntax v0.6.22
    Checking scopeguard v1.1.0
   Compiling log v0.4.14
   Compiling maybe-uninit v2.0.0
   Compiling byteorder v1.3.4
   Compiling getrandom v0.2.0
   Compiling serde_derive v1.0.125
   Compiling bitflags v1.2.1
   Compiling ucd-trie v0.1.3
   Compiling serde v1.0.125
    Checking same-file v1.0.6
   Compiling version_check v0.9.3
   Compiling ryu v1.0.5
    Checking instant v0.1.6
    Checking smallvec v1.6.1
    Checking fnv v1.0.7
   Compiling maplit v1.0.2
    Checking pin-project-lite v0.2.4
   Compiling rustc-rayon-core v0.3.1
   Compiling serde_json v1.0.59
    Checking ppv-lite86 v0.2.8
   Compiling pulldown-cmark v0.8.0
    Checking ansi_term v0.12.1
    Checking itoa v0.4.6
    Checking either v1.6.0
    Checking termcolor v1.1.2
    Checking remove_dir_all v0.5.3
    Checking macro-utils v0.1.3
    Checking arrayvec v0.7.0
    Checking thread_local v1.0.1
    Checking tracing-core v0.1.17
    Checking sharded-slab v0.1.1
    Checking lock_api v0.4.1
   Compiling crossbeam-utils v0.7.2
   Compiling memoffset v0.5.5
   Compiling crossbeam-epoch v0.8.2
   Compiling crossbeam-utils v0.8.3
    Checking walkdir v2.3.1
   Compiling pest v2.1.3
   Compiling unicase v2.6.0
    Checking itertools v0.9.0
    Checking minifier v0.0.41
    Checking tracing-log v0.1.2
    Checking aho-corasick v0.7.13
    Checking bstr v0.2.13
   Compiling quote v1.0.7
    Checking regex-automata v0.1.9
   Compiling pest_meta v2.1.3
    Checking parking_lot_core v0.8.3
    Checking num_cpus v1.13.0
    Checking atty v0.2.14
    Checking crossbeam-queue v0.2.3
    Checking rand_core v0.6.2
    Checking parking_lot v0.11.1
    Checking rand_chacha v0.3.0
    Checking regex v1.4.3
    Checking crossbeam-deque v0.7.3
    Checking rand v0.8.3
    Checking matchers v0.0.1
    Checking rustc-rayon v0.3.1
    Checking tempfile v3.2.0
    Checking globset v0.4.5
    Checking ignore v0.4.17
    Checking globwalk v0.8.1
   Compiling pest_generator v2.1.3
   Compiling tracing-attributes v0.1.13
   Compiling pest_derive v2.1.0
    Checking tracing v0.1.25
    Checking tracing-subscriber v0.2.16
    Checking tracing-tree v0.1.9
    Checking rustdoc-json-types v0.1.0 (/home/joshua/rustc4/src/rustdoc-json-types)
    Checking tera v1.10.0
    Checking rustdoc v0.0.0 (/home/joshua/rustc4/src/librustdoc)
error[E0432]: unresolved import `x`
  --> src/librustdoc/lib.rs:20:5
   |
20 | use x;
   |     ^ no external crate `x`

error: aborting due to previous error

For more information about this error, try `rustc --explain E0432`.
error: could not compile `rustdoc`

To learn more, run the command again with --verbose.
Build completed unsuccessfully in 0:00:18
```

</details>

I agree the atexit stuff is overkill, but I do think it's more useful to show at the end (showing it before seems fine to do in addition if you like). What do you think about making that part of the of destructor for `Step` so it runs automatically?

**Comment by jyn514** [comment]
@Mark-Simulacrum I saw you switched the labels back and forth - did you intend for them to end up as `waiting-on-author`? Happy to implement `step()` and the other implementation details in the meantime, but I'd like to hear a response to https://github.com/rust-lang/rust/pull/86022#issuecomment-873749934 before merging.

**Comment by jyn514** [comment]
> I think an alternative to this, which may be more broadly useful, is to change the formatting of the "step start" messages we currently emit such that they contain a more close reference to the necessary command to invoke that step, should any paths be available.

I implemented this. I'm feeling a little burned out on this PR, I'd like to land what I have so far if at all possible.

**Comment by JohnCSimon** [comment]
Ping from triage: 
@jyn514 Merge conflicts

**Pull request closed without merging**